### PR TITLE
fix: Time shifts calculation for ECharts plugins

### PR DESCRIPTION
**Pull request by michael-s-molina**
### SUMMARY
As part of of [[SIP-50] Proposal for using ECharts as our main charting library](https://github.com/apache/superset/issues/10418), we're currently testing Line and Area chart migrations using Airbnb's production data. During these tests, we found many problems when comparing how legacy and ECharts plugins calculate and display time shifts. This PR fixes the following issues:

### Time shifts when the their granularity do not match the chart's granularity

In this example, we have a chart with a monthly granularity but with a 52 weeks offset. This is how the legacy line chart correctly renders the chart:
Screenshot 2024-05-10 at 14 35 24
<br>The ECharts line chart was incorrectly calculating the time shift when applying the 52 weeks offset due to differences in terms of granularity. Observe how the peak in November 2003 (296) does not match the time shift in 2004 (70).
Screenshot 2024-05-10 at 14 57 56
<br>Now the ECharts line chart correctly calculates the time shift and matches its legacy version.
Screenshot 2024-05-10 at 14 35 52

### How time shifts are displayed when there's missing data
Previously, when time shift data was not available, the chart was rendered with line gaps which is different than the legacy version. 
Screenshot 2024-05-10 at 14 37 19
<br>Now, the line gaps are gone and the dots are connected without a data point which is exactly how the legacy chart displays this scenario.
Screenshot 2024-05-10 at 14 35 52
<br>There is also an error about how the legacy version displays missing data depending on the chosen granularity. In this example we have a chart with daily granularity and we only have time shift data for some data points. The legacy chart displays a line for every data point and assumes that the time shift is the same as the series which is not correct.
Screenshot 2024-05-10 at 14 32 53
<br>With this fix, only available time shifts are displayed and the tooltip display the correct information when a time shift is not available.
Screenshot 2024-05-10 at 14 33 37

### How multiple time shifts are displayed
Previously, multiple time shifts were displayed using the same line pattern, which is confusing and require users to check the tooltips to see the actual time shift.
Screenshot 2024-05-10 at 14 52 42
<br>Now, it behaves exactly as the legacy version where each time shift is displayed using a different line pattern.
Screenshot 2024-05-10 at 14 51 19

### TESTING INSTRUCTIONS
To test the PR use the Advanced Analytics controls and play with different time shift settings. I tested the algorithm using Airbnb's production data using more than 100 charts and added extra unit tests.

### ADDITIONAL INFORMATION
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by john-bodley — COMMENTED** [review]
Thanks @michael-s-molina for the fix. Overall this LGTM, especially given we synced up offline to discuss the underlying logic.

File: superset/common/query_context_processor.py [hunk 1]
```diff
@@ -519,51 +510,112 @@ def processing_time_offsets(  # pylint: disable=too-many-locals,too-many-stateme
...
+            # left join df with offset_df
+            df = dataframe_utils.left_join_df(
+                left_df=df,
+                right_df=offset_df,
+                join_keys=actual_join_keys,
+                rsuffix=R_SUFFIX,
+            )
+
+            # move the temporal column to the first column in df
+            col = df.pop(join_keys[0])
```
[the remark this answers is not in the document]
Comment (michael-s-molina) [reply superset/common/query_context_processor.py:572]: Yep

File: tests/unit_tests/common/test_time_shifts.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+
+    result = query_context_processor.join_offset_dfs(
+        df, offset_dfs, time_grain, join_keys
+    )
+
+    assert_frame_equal(df, result)
+
+
+def test_join_offset_dfs_with_offsets():
+    df = DataFrame({"A": ["2021-01-01", "2021-02-01", "2021-03-01"]})
```
[the remark this answers is not in the document]
Comment (michael-s-molina) [reply tests/unit_tests/common/test_time_shifts.py:91]: Their content depends on the test case for most of them.

File: superset/common/query_context_processor.py [hunk 3]
```diff
@@ -339,19 +339,31 @@ def get_time_grain(query_object: QueryObject) -> Any | None:
...
+        """
+        Adds an offset join column to the provided DataFrame.
+
+        :param df: pandas DataFrame to which the offset join column will be added.
+        :param name: The name of the new column to be added.
+        :param time_grain: The time grain used to calculate the new column.
+        :param time_offset: The time offset used to calculate the new column.
+        :param join_column_producer: A function to generate the join column.
+
+        :return: None. The function modifies the DataFrame in-place.
```
[the remark this answers is not in the document]
Comment (michael-s-molina) [reply superset/common/query_context_processor.py]: Done!

File: superset/common/query_context_processor.py [hunk 4]
```diff
@@ -339,19 +339,31 @@ def get_time_grain(query_object: QueryObject) -> Any | None:
...
         self,
         df: pd.DataFrame,
+        name: str,
         time_grain: str,
+        time_offset: str | None = None,
         join_column_producer: Any = None,
     ) -> None:
+        """
+        Adds an offset join column to the provided DataFrame.
+
```
[the remark this answers is not in the document]
Comment (michael-s-molina) [reply superset/common/query_context_processor.py:353]: Done!

File: superset/common/query_context_processor.py [hunk 5]
```diff
@@ -519,51 +510,112 @@ def processing_time_offsets(  # pylint: disable=too-many-locals,too-many-stateme
...
+        for offset, offset_df in offset_dfs.items():
+            # defines a column name for the offset join column
+            column_name = OFFSET_JOIN_COLUMN_SUFFIX + offset
+
+            # add offset join column to df
+            self.add_offset_join_column(
+                df, column_name, time_grain, offset, join_column_producer
+            )
+
+            # add artifoffseticial join column to offset_df
```
[the remark this answers is not in the document]
Comment (michael-s-molina) [reply superset/common/query_context_processor.py]: Done!

**Review by Antonio-RiveroMartnez — APPROVED** [review]
LGTM!

**Review by eschutho — APPROVED** [review]
lgtm

**Pull request merged**